streams/runners/storm/sampling/src/main/java/com/hortonworks/streamline/streams/sampling/service/storm/StormTopologySamplingService.java

```diff
@@ -0,0 +1,257 @@
+        return buildSamplingStatus(client.getSamplingStatus(topologyId, asUser));
+    }
+
+    @Override
+    public SamplingStatus getSamplingStatus(Topology topology, TopologyComponent component, String asUser) {
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        return buildSamplingStatus(client.getSamplingStatus(topologyId, asUser));
+    }
+
+    @Override
+    public SampledEvents getSampledEvents(Topology topology, TopologyComponent component, EventQueryParams qps, String asUser) {
+        List<SampledEvent> events = new ArrayList<>(qps.count());
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        String componentId = component.getId() + "-" + component.getName();
+        Long cur;
+        if (qps.desc()) {
+            cur = qps.start();
+        } else {
+            cur = qps.start() == null ? Long.valueOf(0) : qps.start();
+        }
+        Long nextOffset;
+        Long prevOffset = null;
+        do {
+            String res = client.getSampledEvents(topologyId, componentId, asUser, cur, qps.length());
+            Matcher match = LOG_CONTENT_REGEX.matcher(res);
+            if (!match.find()) {
+                break;
+            }
+            String lines = match.group(1);
+            if (!StringUtils.isEmpty(lines)) {
+                addSampledEvents(lines, componentId, events);
+            }
+            match = PREV_START_REGEX.matcher(res);
+            prevOffset = match.find() ? Long.parseLong(match.group(1)) : null;
+            match = NEXT_START_REGEX.matcher(res);
+            nextOffset = match.find() ? Long.parseLong(match.group(1)) : null;
+            cur = qps.desc() ? prevOffset : nextOffset;
+        } while (events.size() < qps.count() && cur != null);
+
+        long next = 0;
+        Integer length = null;
+        if (!events.isEmpty()) {
+            if (events.size() > qps.count()) {
+                int i = qps.desc() ? events.size() - qps.count() : 0;
+                int j = qps.desc() ? events.size() : qps.count();
+                events = events.subList(i, j);
+            }
+            if (qps.desc()) {
+                if (qps.start() != null) {
+                    next = qps.start() + events.get(0).getStartOffset() - qps.length() + 1;
+                } else if (prevOffset != null) {
+                    next = prevOffset + events.get(0).getStartOffset() + 1;
+                }
+                if (next >= 0) {
+                    length = qps.length();
+                } else if (qps.start() > 0 && qps.start() < qps.length()){
+                    length = events.get(0).getStartOffset() + qps.start().intValue() + 1;
+                } else {
+                    length = events.get(0).getStartOffset() + 1;
+                }
+                next = Math.max(next, 0);
+                Collections.reverse(events);
```

Comment:
This will not work with multiple pages. 

Suppose first page contains [1, 2, 3, 4, 5] and second page contains [6, 7, 8, 9, 10]. We're fetching page in reverse order so events will be [6, 7, 8, 9, 10, 1, 2, 3, 4, 5] and reversed events will be [5, 4, 3, 2, 1, 10, 9, 8, 7, 6].

Simple fix is adding events per page by reversed order (if `desc()` is true), so that events will be [10, 9, 8, 7, 6] for the first fetch, and [10, 9, 8, 7, 6, 5, 4, 3, 2, 1] for the second fetch.

Reply:
Nice catch. Will fix.